lib/rex/proto/dns/resolver.rb

```diff
@@ -375,8 +378,99 @@ def query(name, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
       return send(name,type,cls)
 
     end
+  end # Resolver
+
+  class CachedResolver < Resolver
+    include Rex::Proto::DNS::Constants
+    attr_accessor :cache
 
+    #
+    # Initialize resolver with cache
+    #
+    # @param config [Hash] Resolver config
+    #
+    # @return nil
+    def initialize(config = {})
+      super(config)
+      self.cache = Rex::Proto::DNS::Cache.new
+      # Read hostsfile into cache
+      hf = Rex::Compat.is_windows ? '%WINDIR%/system32/drivers/etc/hosts' : '/etc/hosts'
+      entries = File.read(hf).lines.map(&:strip).select do |entry|
```

Comment:
My presumption is that since we use privileged functions like raw sockets, we're running privileged.
Can wrap 398 in a BRE if you think that to be a worth-while safety latch

Reply:
How's this?

Reply:
That looks good, nice work!